pandas/io/rdata.py

```diff
@@ -0,0 +1,588 @@
+    ...                  'Goddard',
+    ...                  'Goddard',
+    ...                  'Goddard',
+    ...                  'NRTSI-G'),
+    ...     region = c('S', 'S', 'S', 'S', 'S'),
+    ...     extent = c(8.28, 9.48, 9.19, 9.41, 10.44),
+    ...     area = c(5.51, 6.23, 5.59, 6.59, 6.5)
+    ... )
+    ... save(ghg_df, plants_df, sea_ice_df, file="env_data_dfs.rda")
+    ... '''
+
+    >>> env_dfs = pd.read_rdata("env_data_dfs.rda")  # doctest: +SKIP
+    >>> env_dfs  # doctest: +SKIP
+    {{'ghg_df':
+                          gas  year  emissions
+    rownames
+    1          Carbon dioxide  2018    5424.88
+    2                 Methane  2018     634.46
+    3           Nitrous oxide  2018     434.53
+    4       Fluorinated gases  2018     182.79
+    5                   Total  2018    6676.65,
+    'plants_df':
+               plant_group               status  count
+    rownames
+    1        Pteridophytes       Data Deficient    398
+    2        Pteridophytes              Extinct     65
+    3        Pteridophytes       Not Threatened   1294
+    4        Pteridophytes  Possibly Threatened    408
+    5        Pteridophytes           Threatened   1275,
+    'sea_ice_df':
+           year  mo data.type region  extent  area
+    rownames
+    1      2016  12   Goddard      S    8.28  5.51
+    2      2017  12   Goddard      S    9.48  6.23
+    3      2018  12   Goddard      S    9.19  5.59
+    4      2019  12   Goddard      S    9.41  6.59
+    5      2020  12   NRTSI-G      S   10.44  6.50}}
+    """
+
+    import_optional_dependency("pyreadr")
+
+    rdr = _PyReadrParser(
+        path_or_buffer,
+        file_format,
+        select_frames,
+        rownames,
+        storage_options,
+    )
+
+    return rdr.parse_data()
+
+
+def _get_data_from_filepath(
+    filepath_or_buffer,
+    encoding,
+    compression,
+    storage_options,
+) -> Union[str, bytes, Buffer]:
+    """
+    Extract raw R data.
+
+    The method accepts three input types:
+        1. filepath (string-like)
+        2. file-like object (e.g. open file object, BytesIO)
+        3. R data file in ascii or binary content
+
+    This method turns (1) into (2) to simplify the rest of the processing.
+    It returns input types (2) and (3) unchanged.
+    """
+    filepath_or_buffer = stringify_path(filepath_or_buffer)
+
+    if (
+        not isinstance(filepath_or_buffer, str)
+        or is_url(filepath_or_buffer)
+        or is_fsspec_url(filepath_or_buffer)
+        or file_exists(filepath_or_buffer)
+    ):
+        with get_handle(
+            filepath_or_buffer,
+            "rb",
+            encoding=encoding,
+            compression=compression,
+            storage_options=storage_options,
+            is_text=False,
+        ) as handle_obj:
+            filepath_or_buffer = (
+                handle_obj.handle.read()
+                if hasattr(handle_obj.handle, "read")
+                else handle_obj.handle
```

Comment:
When is `handle_obj.handle` used and does it return the content?

Reply:
IIUC, all valid file/URL paths and buffers render in `handle_obj.handle`. While buffers return with `read()`, stringified paths return unchanged. I have tests for both file and file-like types:

```python
def test_read_rds_file(datapath):
    filename = datapath("io", "data", "rdata", "ghg_df.rds")
    r_df = read_rdata(filename)
    ...

def test_bytes_read_rda(datapath):
    filename = datapath("io", "data", "rdata", "env_data_dfs.rda")

    with open(filename, "rb") as f:
        r_dfs = read_rdata(f, file_format="rda")
    ...

def test_bytesio_rds(datapath):
    filename = datapath("io", "data", "rdata", "sea_ice_df.rds")

    with open(filename, "rb") as f:
        with BytesIO(f.read()) as b_io:
            r_df = read_rdata(b_io, file_format="rds")
    ...
```

R data files can also have an ASCII (i.e., text) version format (see [docs](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/save)), so I accommodated this as well with different modes within code. 

```python
with _preprocess_data(handle_data) as r_data:
    mode = "wb" if isinstance(r_data, io.BytesIO) else "w"
    with open(r_temp, mode) as f:
        f.write(r_data.read())
````